Monika After Story/game/definitions.rpy

```diff
@@ -28,6 +29,108 @@ python early:
     renpy.execution.check_infinite_loop = dummy
 
+    class MASFormatter(renpy.substitutions.Formatter):
+        """
+        Our string formatter that uses more
+        advanced formatting rules compared to the RenPy one
+        """
+        def get_field(self, field_name, args, kwargs):
+            """
+            Originally this method returns objects by references
+            Our variant allows us to eval functions, e.g. "Text [my_func(arg1)]."
+            and use negative indexes for iterables, e.g. "Text [my_iterable[-2]]."
+
+            IN:
+                field_name - the reference to the object
+                args - not sure, but renpy doesn't use it (passes in an empty tuple)
+                kwargs - the store modules where renpy will look for the object
+
+            OUT:
+                tuple of the object and its key
+            """
+            def _getStoreNameForObject(object_name, *scopes):
+                """
+                Returns the name of the store where the given object
+                was defined or imported to
+
+                IN:
+                    object_name - the name of the object to look for (string)
+                    scopes - the scopes where we look for the object (storemodule.__dict__)
+
+                OUT:
+                    name of the store module where the object was defined
+                    or empty string if we couldn't find it
+                """
+                for scope in scopes:
+                    if object_name in scope:
+                        stores_names_list = [
+                            store_module_name
+                            for store_module_name, store_module in sys.modules.iteritems()
+                            if store_module and store_module.__dict__ is scope
+                        ]
+                        if stores_names_list:
+                            return stores_names_list[0]
+
+                return ""
+
+            # if it's a function call, we eval it
+            if "(" in field_name:
+                # split the string into its components
+                func_name, paren, args = field_name.partition("(")
+
+                # it still may include store modules, get the first if there's any
+                if "." in func_name:
+                    func_store_name, dot, func_name = func_name.partition(".")
+                    first = func_store_name
+
+                # otherwise work with just function
+                else:
+                    func_store_name = ""
+                    dot = ""
+                    first = func_name
+
+                # now we find the store's name to use in eval
+                if isinstance(kwargs, renpy.substitutions.MultipleDict):
+                    scope_store_name = _getStoreNameForObject(first, *kwargs.dicts)
+
+                else:
+                    scope_store_name = _getStoreNameForObject(first, kwargs)
```

Comment:
if partition doesn't find the string, it returns `("string you called it on", "", "")`, so you can drop the if here.

```suggestion
                # it still may include store modules, get the first if there's any
                func_store_name, dot, func_name = func_name.partition(".")

                # now we find the store's name to use in eval
                if isinstance(kwargs, renpy.substitutions.MultipleDict):
                    scope_store_name = _getStoreNameForObject(func_store_name, *kwargs.dicts)

                else:
                    scope_store_name = _getStoreNameForObject(func_store_name, kwargs)
```

Reply:
I still need to return `first`, the OG function does so. But I can get rid of the dupe code.